File: rebuild-shop.js
const fs = require('fs');
const path = require('path');

const shopPath = path.join(__dirname, 'shop.html');
const current = fs.readFileSync(shopPath, 'utf8');
const newBooks = JSON.parse(fs.readFileSync(path.join(__dirname, 'books-100.json'), 'utf8'));

// Extract existing 16 books from current shop.html
const startIdx = current.indexOf('const PRODUCTS = [');
const endIdx = current.indexOf('];\n\nlet activeFilter');
const existingBlock = current.slice(startIdx, endIdx + 2);
const existingMatch = existingBlock.match(/const PRODUCTS = ([\s\S]*?)\];/);
const existing = new Function('return ' + existingMatch[1] + ']')();
const orig16 = existing.slice(0, 16);

console.log('Original books:', orig16.length);
console.log('New books:', newBooks.length);

const allBooks = [...orig16, ...newBooks];
console.log('Total:', allBooks.length);

function esc(s) {
  return s.replace(/\\/g, '\\\\').replace(/"/g, '\\"').replace(/\n/g, '\\n');
}

const prodStr = 'const PRODUCTS = [\n' + allBooks.map(b => {
  const fields = [];
  fields.push(`id:${b.id}`);
  fields.push(`title:"${esc(b.title)}"`);
  fields.push(`author:"${esc(b.author)}"`);
  fields.push(`year:${b.year}`);
  fields.push(`cat:"${b.cat}"`);
  fields.push(`price:${b.price.toFixed(2)}`);
  if (b.coverImage) fields.push(`coverImage:"${b.coverImage}"`);
  if (b.freeLink) fields.push(`freeLink:"${b.freeLink}"`);
  fields.push(`buyLink:"${b.buyLink}"`);
  fields.push(`blurb:"${esc(b.blurb)}"`);
  fields.push(`reason:"${esc(b.reason)}"`);
  return '  { ' + fields.join(', ') + ' }';
}).join(',\n') + '\n];';

let newHtml = current.slice(0, startIdx) + prodStr + current.slice(endIdx + 2);
newHtml = newHtml.replace(/16 forbidden books/, '116 forbidden books');

fs.writeFileSync(shopPath, newHtml);
console.log('shop.html rebuilt successfully');

File: rebuild-shop.py
import json, re

with open('/Users/openclaw/Desktop/WBF/shop.html') as f:
    html = f.read()

with open('/Users/openclaw/Desktop/WBF/books-100.json') as f:
    new_books = json.load(f)

# Extract existing 16 books blocks (IDs 1-16) from current html
prod_section = html.split('const PRODUCTS =')[1].split('];')[0]
existing_blocks = []
for m in re.finditer(r'\{[^}]*id:(\d+)[^}]*\}', prod_section):
    bid = int(m.group(1))
    if 1 <= bid <= 16:
        existing_blocks.append((bid, m.group(0)))

existing_blocks.sort(key=lambda x: x[0])
print(f'Extracted {len(existing_blocks)} original books')

# Build fresh array
all_books = [b[1] for b in existing_blocks]

for b in new_books:
    fields = []
    fields.append(f'id:{b["id"]}')
    fields.append(f'title:"{b["title"].replace(chr(92), chr(92)+chr(92)).replace(chr(34), chr(92)+chr(34))}"')
    fields.append(f'author:"{b["author"].replace(chr(92), chr(92)+chr(92)).replace(chr(34), chr(92)+chr(34))}"')
    fields.append(f'year:{b["year"]}')
    fields.append(f'cat:"{b["cat"]}"')
    fields.append(f'price:{b["price"]:.2f}')
    if b.get('coverImage'):
        fields.append(f'coverImage:"{b["coverImage"]}"')
    if b.get('freeLink'):
        fields.append(f'freeLink:"{b["freeLink"]}"')
    fields.append(f'buyLink:"{b["buyLink"]}"')
    fields.append(f'blurb:"{b["blurb"].replace(chr(92), chr(92)+chr(92)).replace(chr(34), chr(92)+chr(34))}"')
    fields.append(f'reason:"{b["reason"].replace(chr(92), chr(92)+chr(92)).replace(chr(34), chr(92)+chr(34))}"')
    all_books.append('  { ' + ', '.join(fields) + ' }')

prod_str = 'const PRODUCTS = [\n' + ',\n'.join(all_books) + '\n];'

# Replace the old array
start = html.index('const PRODUCTS = [')
end = html.index('];\n\nlet activeFilter')
if end == -1:
    end = html.index('];\n\nfunction renderGrid')
new_html = html[:start] + prod_str + html[end+2:]

# Update header
new_html = new_html.replace('16 forbidden books', '116 forbidden books')

with open('/Users/openclaw/Desktop/WBF/shop.html', 'w') as f:
    f.write(new_html)

print('shop.html rebuilt successfully')

# Verify
with open('/Users/openclaw/Desktop/WBF/shop.html') as f:
    verify = f.read()
ids = re.findall(r'id:(\d+)', verify.split('const PRODUCTS =')[1].split('];')[0])
print(f'Total IDs: {len(ids)}')
new_with_cover = sum(1 for m in re.finditer(r'id:(\d+).*?coverImage', verify.split('const PRODUCTS =')[1].split('];')[0]) if int(m.group(1)) >= 17)
print(f'New books with covers: {new_with_cover}/100')
